### PR TITLE
WT-4642 Store transaction IDs durably.

### PR DESCRIPTION
**Pull request by keithbostic**
This is the first set of changes for WT-4642, adding start/stop transactions to the on-leaf-page values. I think it's straight-forward, things to note:

* I went with the generic word "history" to describe what I'm putting on the page, that is, a set of timestamps and/or transaction IDs are the value's "history".

* I added `WT_TXN_MAX` because comparing a stop transaction to `WT_TXN_ABORTED` seems like it would be difficult to read. I used:
```
#define WT_TXN_MAX      UINT64_MAX      /* End of time */
#define WT_TXN_ABORTED  (UINT64_MAX - 1)/* Update rolled back, ignore. */
```

* @michaelcahill, I changed the `upd_select` work in `__rec_upd_select()`, and note that it's not modifying the actual `WT_UPDATE` structure.

Let me know if you disagree with these decisions.

### CODE REVIEW

**Comment by keithbostic** [comment]
3fd41e2 is the second part of this change, updating the internal page address cells.

**Comment by keithbostic** [comment]
autobuild: retest this please

**Comment by keithbostic** [comment]
autobuild: retest this please

**Comment by keithbostic** [comment]
@agorrod, @sulabhM, @michaelcahill, I think this set of changes is complete and should be ready for review & merge.

File: src/include/cell.i [hunk 1]
```diff
@@ -369,11 +487,19 @@ __wt_cell_pack_value_match(WT_CELL *page_cell,
 		++a;
 	} else if (WT_CELL_TYPE(a[0]) == WT_CELL_VALUE) {
 		rle = (a[0] & WT_CELL_64V) != 0;
-		ts = (a[0] & WT_CELL_TIMESTAMPS) != 0;
+		validity = (a[0] & WT_CELL_SECOND_DESC) != 0;
 		++a;
-		if (ts) {
-			WT_RET(__wt_vunpack_uint(&a, 0, &v));	/* Skip TS */
-			WT_RET(__wt_vunpack_uint(&a, 0, &v));
+		if (validity) {			/* Skip validity window */
```
Comment (agorrod) [inline src/include/cell.i]: Having to skip like this makes me a bit sad - though I can't think of a better solution. We can't really move where the validity window sits in the data structure, and we don't want to pay the cost of saving the size of the validity window into the cell.

File: src/btree/bt_handle.c [hunk 2]
```diff
@@ -121,7 +121,6 @@ __wt_btree_open(WT_SESSION_IMPL *session, const char *op_cfg[])
 		F_SET(btree, WT_BTREE_READONLY);
 
 	/* Get the checkpoint information for this name/checkpoint pair. */
-	WT_CLEAR(ckpt);
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/btree/bt_handle.c]: There were two callers of `__wt_meta_checkpoint()`, and both of them cleared the memory before the call, I thought it cleaner to have `__wt_meta_checkpoint()` do the clear instead of the callers.

File: src/btree/bt_page.c [hunk 3]
```diff
@@ -346,7 +369,14 @@ __inmem_col_var(
 	 */
 	indx = 0;
 	cip = page->pg_var;
-	WT_CELL_FOREACH_BEGIN(session, btree, page->dsk, unpack, true) {
+	WT_CELL_FOREACH_BEGIN(session, btree, page->dsk, unpack) {
+		/* Optionally skip unstable values */
+		if (check_unstable &&
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/btree/bt_page.c]: > Could you push the `check_unstable` parameter inside `__unstable_skip` and also pass page directly and do the decrement of `entries` in the function? That would reduce the code duplication amongst callers of `__unstable_skip`

I left the code this way for two reasons. First, I think `__unstable_skip()` may become more complicated in the future (for example, if we were to start skipping address blocks in internal pages for some reason), and I'd prefer to keep its functionality strictly limited for now. Second, I'm unhappy with the decrement of `WT_PAGE.entries`, I'm concerned the inaccuracy this implies in page counts isn't OK and it will need adjustment, in which case those fixes would have to be in the code above this function, not `__unstable_skip()`.

In summary, the structure was deliberate, but I don't have particularly good reasons for the decision  -- if you feel strongly, happy to make the change.

File: src/btree/bt_split.c [hunk 4]
```diff
@@ -1448,7 +1450,7 @@ __split_multi_inmem(
 	 * allocated page on error, when discarding the allocated WT_REF.
 	 */
 	WT_RET(__wt_page_inmem(
-	    session, ref, multi->disk_image, WT_PAGE_DISK_ALLOC, &page));
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/btree/bt_split.c]: > I'm surprised the split code is ignoring timestamp/txnid information - was that deliberate?

What we're doing here is re-instantiating a page from a disk image the reconciliation code has just created, and reconciliation won't create a disk image with entries the running database no longer cares about, at least, not based on the tests we're currently performing in `__unstable_skip()`, so there wasn't a reason to  pay the performance cost.

In the future, we might add visibility tests to `__unstable_skip()`, and then the boolean might toggle?

File: src/include/cell.i (hunk 1 above)
Comment (keithbostic) [reply src/include/cell.i]: > Having to skip like this makes me a bit sad - though I can't think of a better solution. We can't really move where the validity window sits in the data structure, and we don't want to pay the cost of saving the size of the validity window into the cell.

We could move where the validity window sits in the cell. I started out putting it at the end of the cell after the data chunk because that was an easier code change, but .eventually decided that was harder to think about, so I went with what's there now. If you feel strongly, this could change.

File: src/include/cell.i [hunk 5]
```diff
@@ -17,8 +17,9 @@
  * There are 4 basic cell types: keys and data (each of which has an overflow
  * form), deleted cells and off-page references.  The cell is usually followed
  * by additional data, varying by type: keys are followed by a chunk of data,
- * data is followed by optional timestamps and a chunk of data, overflow and
- * off-page cells are followed by optional timestamps and an address cookie.
+ * values are is followed by an optional validity window and a chunk of data,
```
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/include/cell.i]: Fixed, pushed in 146acb7.

File: src/btree/bt_handle.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (agorrod) [reply src/btree/bt_handle.c]: ack

File: src/btree/bt_page.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (agorrod) [reply src/btree/bt_page.c]: I don't feel strongly - since you had a reason I'll resolve.

File: src/btree/bt_split.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (agorrod) [reply src/btree/bt_split.c]: That makes sense - I did read the existing comment but it still wasn't clear to me, would you mind updating the comment to make it clear why it's safe to skip unstable updates?

File: src/include/cell.i (hunk 1 above)
Comment (agorrod) [reply src/include/cell.i]: Overall I think it's better where you have added it.

File: src/btree/bt_split.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/btree/bt_split.c]: Fixed, pushed in 684e384.

**Comment by michaelcahill** [comment]
Thanks @keithbostic, I've reviewed these changes and they lgtm.  I'll merge...
